File: variations-stuck-on-backorder-at-zero/node/fix-variation-stock-status.js
/**
 * Repair WooCommerce variations whose stock_status disagrees with their stock_quantity.
 *
 * A variation can end up showing "On backorder" in the shop while its stock is at or
 * below zero and backorders are turned off. WooCommerce only recalculates
 * stock_status when the quantity changes through its own save path. A CSV import, a
 * direct database edit, or flipping the backorders setting after the quantity was
 * already low can leave the stored stock_status stale. This walks the variations of
 * a product (or every variable product), works out what stock_status should be from
 * the quantity and the backorders setting, and corrects any variation that
 * disagrees. Read only by default. Run it by hand or on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/variations-stuck-on-backorder-at-zero/
 */
import { pathToFileURL } from "node:url";

const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const PRODUCT_IDS = (process.env.PRODUCT_IDS || "").split(",").map((s) => s.trim()).filter(Boolean);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const VALID_STATUSES = new Set(["instock", "outofstock", "onbackorder"]);

/**
 * Work out the stock_status a variation should have.
 *
 * Only variations with manage_stock on carry their own quantity, so anything
 * else is left to WooCommerce and skipped. Backorders "yes" or "notify" both
 * mean the shop should keep selling once stock runs out.
 */
export function expectedStockStatus(variation) {
  if (!variation.manage_stock) return null;
  const qty = variation.stock_quantity;
  if (qty === null || qty === undefined) return null;
  const backorders = variation.backorders || "no";
  if (qty > 0) return "instock";
  if (backorders === "yes" || backorders === "notify") return "onbackorder";
  return "outofstock";
}

/**
 * Pure decision: does this variation's stock_status need to change.
 * Returns ["skip" | "fix", reason]. No I/O happens in here, so it is safe
 * and cheap to unit test.
 */
export function decide(variation) {
  const expected = expectedStockStatus(variation);
  if (expected === null) return ["skip", "variation does not manage its own stock"];
  const current = variation.stock_status;
  if (!VALID_STATUSES.has(current)) {
    return ["fix", `stock_status ${JSON.stringify(current)} is not a recognized value`];
  }
  if (current === expected) return ["skip", "stock_status already matches quantity and backorders"];
  return ["fix", `stock_status is ${JSON.stringify(current)} but should be ${JSON.stringify(expected)}`];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* listVariableProducts() {
  let page = 1;
  while (true) {
    const batch = await woo(`/products?type=variable&per_page=50&page=${page}&status=publish`);
    if (!batch.length) return;
    for (const product of batch) yield product;
    page++;
  }
}

async function* listVariations(productId) {
  let page = 1;
  while (true) {
    const batch = await woo(`/products/${productId}/variations?per_page=100&page=${page}`);
    if (!batch.length) return;
    for (const variation of batch) yield variation;
    page++;
  }
}

async function applyFix(productId, variationId, expectedStatus) {
  await woo(`/products/${productId}/variations/${variationId}`, {
    method: "PUT",
    body: JSON.stringify({ stock_status: expectedStatus }),
  });
}

async function targetProductIds() {
  if (PRODUCT_IDS.length) return PRODUCT_IDS;
  const ids = [];
  for await (const product of listVariableProducts()) ids.push(product.id);
  return ids;
}

export async function run() {
  let fixed = 0;
  for (const productId of await targetProductIds()) {
    for await (const variation of listVariations(productId)) {
      const [action, reason] = decide(variation);
      if (action === "skip") continue;
      const expected = expectedStockStatus(variation);
      console.log(
        `Variation ${variation.id} (product ${productId}): ${reason}. ${DRY_RUN ? "would fix" : "fixing"}`
      );
      if (!DRY_RUN) await applyFix(productId, variation.id, expected);
      fixed++;
    }
  }
  console.log(`Done. ${fixed} variation(s) ${DRY_RUN ? "to fix" : "fixed"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: variations-stuck-on-backorder-at-zero/python/fix_variation_stock_status.py
"""Repair WooCommerce variations whose stock_status disagrees with their stock_quantity.

A variation can end up showing "On backorder" in the shop while its stock is at or
below zero and backorders are turned off. WooCommerce only recalculates
stock_status when the quantity changes through its own save path. A CSV import, a
direct database edit, or flipping the backorders setting after the quantity was
already low can leave the stored stock_status stale. This walks the variations of a
product (or every variable product), works out what stock_status should be from the
quantity and the backorders setting, and corrects any variation that disagrees.
Read only by default. Run it by hand or on a schedule.
"""
import os
import logging
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_variation_stock_status")

WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
PRODUCT_IDS = [p.strip() for p in os.environ.get("PRODUCT_IDS", "").split(",") if p.strip()]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

VALID_STATUSES = {"instock", "outofstock", "onbackorder"}


def expected_stock_status(variation):
    """Work out the stock_status a variation should have.

    Only variations with manage_stock on carry their own quantity, so anything
    else is left to WooCommerce and skipped. Backorders "yes" or "notify" both
    mean the shop should keep selling once stock runs out.
    """
    if not variation.get("manage_stock"):
        return None
    qty = variation.get("stock_quantity")
    if qty is None:
        return None
    backorders = variation.get("backorders", "no")
    if qty > 0:
        return "instock"
    if backorders in ("yes", "notify"):
        return "onbackorder"
    return "outofstock"


def decide(variation):
    """Pure decision: does this variation's stock_status need to change.

    Returns a tuple of (action, reason). action is one of:
      "skip"  - not stock managed, or already correct
      "fix"   - stock_status disagrees with quantity and backorders, repair it
    No I/O happens in here, so it is safe and cheap to unit test.
    """
    expected = expected_stock_status(variation)
    if expected is None:
        return ("skip", "variation does not manage its own stock")
    current = variation.get("stock_status")
    if current not in VALID_STATUSES:
        return ("fix", f"stock_status {current!r} is not a recognized value")
    if current == expected:
        return ("skip", "stock_status already matches quantity and backorders")
    return ("fix", f"stock_status is {current!r} but should be {expected!r}")


def list_variable_products():
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/products",
            params={"type": "variable", "per_page": 50, "page": page, "status": "publish"},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for product in batch:
            yield product
        page += 1


def list_variations(product_id):
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/products/{product_id}/variations",
            params={"per_page": 100, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for variation in batch:
            yield variation
        page += 1


def apply_fix(product_id, variation_id, expected_status):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/products/{product_id}/variations/{variation_id}",
        json={"stock_status": expected_status},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def target_product_ids():
    if PRODUCT_IDS:
        return PRODUCT_IDS
    return [p["id"] for p in list_variable_products()]


def run():
    fixed = 0
    for product_id in target_product_ids():
        for variation in list_variations(product_id):
            action, reason = decide(variation)
            if action == "skip":
                continue
            expected = expected_stock_status(variation)
            log.info(
                "Variation %s (product %s): %s. %s",
                variation["id"], product_id, reason, "would fix" if DRY_RUN else "fixing",
            )
            if not DRY_RUN:
                apply_fix(product_id, variation["id"], expected)
            fixed += 1
    log.info("Done. %d variation(s) %s.", fixed, "to fix" if DRY_RUN else "fixed")


if __name__ == "__main__":
    run()
